File: embedding_service.py
#!/usr/bin/env python3
"""
本地 Embedding 服务 - BGE-M3
使用 FlagEmbedding，高效、专业、中文支持好
"""

from flask import Flask, request, jsonify
from FlagEmbedding import BGEM3FlagModel
import logging
import time

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask(__name__)

MODEL = None

def init_model():
    global MODEL
    logger.info("🚀 正在加载 BGE-M3 模型...")
    logger.info("   模型: BAAI/bge-m3")
    logger.info("   特点: 中文检索最强，医学场景适用")
    MODEL = BGEM3FlagModel('BAAI/bge-m3', use_fp16=False)
    logger.info("✅ BGE-M3 模型加载完成!")

@app.route('/health', methods=['GET'])
def health():
    return jsonify({"status": "ok", "model": "BGE-M3"})

@app.route('/embed', methods=['POST'])
def embed():
    """单文本 embedding"""
    data = request.json
    text = data.get('text', '')

    if not text:
        return jsonify({"error": "text is required"}), 400

    try:
        start = time.time()
        result = MODEL.encode([text])
        embedding = result['dense_vecs'][0].tolist()
        dimensions = len(embedding)

        return jsonify({
            "embedding": embedding,
            "dimensions": dimensions,
            "model": "BAAI/bge-m3",
            "time_ms": int((time.time() - start) * 1000)
        })
    except Exception as e:
        logger.error(f"Embedding error: {e}")
        return jsonify({"error": str(e)}), 500

@app.route('/embed/batch', methods=['POST'])
def embed_batch():
    """批量文本 embedding"""
    data = request.json
    texts = data.get('texts', [])

    if not texts:
        return jsonify({"error": "texts is required"}), 400

    try:
        start = time.time()
        result = MODEL.encode(texts)
        embeddings = result['dense_vecs'].tolist()

        return jsonify({
            "embeddings": embeddings,
            "count": len(embeddings),
            "dimensions": len(embeddings[0]) if embeddings else 0,
            "model": "BAAI/bge-m3",
            "time_ms": int((time.time() - start) * 1000)
        })
    except Exception as e:
        logger.error(f"Batch embedding error: {e}")
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    init_model()
    app.run(host='127.0.0.1', port=8080, debug=False)

File: src/services/embeddingService.js
const LOCAL_EMBEDDING_URL = 'http://localhost:5000';
const LOCAL_MODEL = 'BGE-small-zh-v1.5';
const EMBEDDING_DIMENSIONS = 512;

export async function initEmbeddingModel() {
    console.log('📥 检查本地 Embedding 服务...');
    console.log(`   模型: ${LOCAL_MODEL}`);
    console.log(`   向量维度: ${EMBEDDING_DIMENSIONS}`);
    console.log(`   服务地址: ${LOCAL_EMBEDDING_URL}`);

    try {
        const response = await fetch(`${LOCAL_EMBEDDING_URL}/health`);
        if (response.ok) {
            console.log('✅ 本地 Embedding 服务已就绪');
            return true;
        } else {
            console.warn(`⚠️ 本地服务响应: ${response.status}`);
            return false;
        }
    } catch (error) {
        console.warn('⚠️ 本地 Embedding 服务不可用:', error.message);
        return false;
    }
}

export async function generateEmbedding(text) {
    try {
        const response = await fetch(`${LOCAL_EMBEDDING_URL}/embed`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ text })
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`本地Embedding错误 ${response.status}: ${error}`);
        }

        const data = await response.json();
        return {
            embedding: data.embedding,
            dimensions: data.dimensions,
            model: data.model
        };
    } catch (error) {
        console.error('Embedding生成失败:', error.message);
        throw error;
    }
}

export async function generateEmbeddingsBatch(texts) {
    try {
        const response = await fetch(`${LOCAL_EMBEDDING_URL}/embed/batch`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ texts })
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`本地Embedding批量错误 ${response.status}: ${error}`);
        }

        const data = await response.json();
        return data.embeddings.map((embedding, i) => ({
            embedding,
            dimensions: embedding.length,
            model: data.model,
            index: i
        }));
    } catch (error) {
        console.error('批量Embedding生成失败:', error.message);
        throw error;
    }
}

export function getEmbeddingDimensions() {
    return EMBEDDING_DIMENSIONS;
}
